src/Microsoft.DotNet.Build.Tasks.Feed/BlobFeedAction.cs

```diff
@@ -0,0 +1,389 @@
+
+                    // Merge index.json involves downloading the existing feed's index.jsons
+                    await Task.WhenAll(items.Select(item => MergeIndexes(CancellationToken, item, relativePath, indexJsonBlobs)));
+                }
+            }
+            if (CancellationToken.IsCancellationRequested)
+            {
+                Log.LogError("Task UploadToAzure cancelled");
+                CancellationToken.ThrowIfCancellationRequested();
+            }
+
+            List<string> indexList = new List<string>(Directory.EnumerateFiles(Path.Combine(IndexDirectory, relativePath), "index.json", SearchOption.AllDirectories));
+            using (var clientThrottle = new SemaphoreSlim(this.MaxClients, this.MaxClients))
+            {
+                Func<string, string> calculatePath = (string item) =>
+                {
+                    Tuple<string, NuGetVersion> blobPath = feed.CalculateBlobPath(item, relativePath);
+                    string relativeBlobPath = Path.Combine(relativePath, blobPath.Item1, blobPath.Item2.ToFullString());
+                    return Path.Combine(relativeBlobPath, Path.GetFileName(item)).ToLowerInvariant();
+                };
+                //upload all items
+                await Task.WhenAll(items.Select(item => UploadAsync(CancellationToken, item, calculatePath(item), clientThrottle)));
+                
+                //upload equivalent index.jsons
+                Func<string, string> calculateIndexPath = (string item) =>
+                {
+                    return item.Substring(IndexDirectory.Length + 1).ToLowerInvariant();
+                };
+                await Task.WhenAll(indexList.Select(item => UploadAsync(CancellationToken, item, calculateIndexPath(item), clientThrottle)));
+            }
+
+            return !Log.HasLoggedErrors;
+        }
+
+        public async Task<bool> PushToFeedFlat(IEnumerable<string> items, string relativePath = "", bool overwrite = false)
+        {
+            if (overwrite)
+            {
+                using (var clientThrottle = new SemaphoreSlim(this.MaxClients, this.MaxClients))
+                {
+                    Func<string, string> calculatePath = (string item) =>
+                    {
+                        return item.Substring(item.IndexOf(relativePath)).Replace("\\", "/");
+                    };
+                    //upload all items
+                    await Task.WhenAll(items.Select(item => UploadAsync(CancellationToken, item, calculatePath(item), clientThrottle)));
+                }
+            }
+            else
+            {
+                throw new Exception("Push failed to due overwrite being false.");
+            }
+            return !Log.HasLoggedErrors;
+        }
+
+        public async Task<List<string>> ListAzureBlobs(string AccountName, string AccountKey, string ContainerName, string FilterBlobNames = "")
+        {
+            List<string> blobsNames = new List<string>();
```

Comment:
We already have a number code paths that do things like list, upload, download, etc. We should be calling and sharing that code instead of duplicating it.

Reply:
fixed.